File: mobile/ios/Sources/ViewModels/LoginViewModel.swift
import Foundation

@MainActor
final class LoginViewModel: ObservableObject {
    @Published var tenantId = ""
    @Published var email = ""
    @Published var password = ""
    @Published var loading = false
    @Published var error: String?

    private let auth = AuthService()
    private let onSignedIn: (Principal) -> Void

    init(onSignedIn: @escaping (Principal) -> Void) {
        self.onSignedIn = onSignedIn
    }

    func submit() {
        guard !email.isEmpty, !password.isEmpty else {
            error = "Email and password are required"
            return
        }
        loading = true
        error = nil
        Task {
            do {
                let principal = try await auth.signIn(
                    email: email,
                    password: password,
                    gcipTenantId: tenantId.isEmpty ? nil : tenantId
                )
                loading = false
                onSignedIn(principal)
            } catch {
                loading = false
                self.error = error.localizedDescription
            }
        }
    }
}

File: mobile/android/app/src/main/kotlin/com/soteria/assurance/ui/login/LoginViewModel.kt
package com.soteria.assurance.ui.login

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.soteria.assurance.data.auth.AuthRepository
import com.soteria.assurance.data.model.Principal
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch

data class LoginUiState(
    val tenantId: String = "",
    val email: String = "",
    val password: String = "",
    val loading: Boolean = false,
    val error: String? = null,
)

class LoginViewModel(
    private val auth: AuthRepository = AuthRepository(),
    private val onSignedIn: (Principal) -> Unit,
) : ViewModel() {

    private val _state = MutableStateFlow(LoginUiState())
    val state: StateFlow<LoginUiState> = _state.asStateFlow()

    fun onTenant(v: String) = _state.update { it.copy(tenantId = v) }
    fun onEmail(v: String) = _state.update { it.copy(email = v) }
    fun onPassword(v: String) = _state.update { it.copy(password = v) }

    fun submit() {
        val s = _state.value
        if (s.email.isBlank() || s.password.isBlank()) {
            _state.update { it.copy(error = "Email and password are required") }
            return
        }
        _state.update { it.copy(loading = true, error = null) }
        viewModelScope.launch {
            runCatching {
                auth.signIn(s.email, s.password, s.tenantId.ifBlank { null })
            }.onSuccess { principal ->
                _state.update { it.copy(loading = false) }
                onSignedIn(principal)
            }.onFailure { e ->
                _state.update { it.copy(loading = false, error = e.message ?: "Sign-in failed") }
            }
        }
    }
}
